src/Polyfill/Polyfill_Task.cs

```diff
@@ -53,23 +54,64 @@ public static Task WaitAsync(
             return Task.FromException(new TimeoutException());
         }
 
-        var delayTask = Task.Delay(timeout, cancellationToken);
+        return WaitAsyncCore(target, timeout, cancellationToken);
+    }
+
+    static async Task WaitAsyncCore(Task task, TimeSpan timeout, CancellationToken cancelToken)
+    {
+        using var timeoutCancelSource = CreateTimeoutCancelSource(timeout, cancelToken, out var combinedToken);
+        var cancelSource = new TaskCompletionSource<bool>();
 
-        Func<Task<Task>, Task> continueFunc = completedTask =>
+        // Register for cancellation/timeout
+        using (combinedToken.Register(() => cancelSource.TrySetCanceled(combinedToken), useSynchronizationContext: false))
```

Comment:
If timeout is specified, and cancellation is requested before the timeout is reached, will we throw OperationCanceledException with the wrong CancellationToken here?